src/array.jl

```diff
@@ -117,10 +113,16 @@ function CategoricalMatrix end
 CategoricalArray(dims::Int...; ordered=false) =
     CategoricalArray{String}(dims, ordered=ordered)
 
-CategoricalArray{T, N, R}(dims::NTuple{N,Int}; ordered=false) where {T, N, R} =
-    CategoricalArray{T, N, R}(zeros(R, dims), CategoricalPool{T, R}(ordered))
+function CategoricalArray{T, N, R}(dims::NTuple{N,Int};
+                                   ordered=false) where {T, N, R}
+    C = catvalue_type(unwrap_catvalue_type(T), R)
+    V = valtype(C)
+    S = T >: Null ? Union{V, Null} : V
+    CategoricalArray{S, N}(zeros(R, dims), CategoricalPool{V, R, C}(ordered))
+end
+
 CategoricalArray{T, N}(dims::NTuple{N,Int}; ordered=false) where {T, N} =
-    CategoricalArray{T, N, DefaultRefType}(dims, ordered=ordered)
+    CategoricalArray{T, N, reftype(T)}(dims, ordered=ordered)
```

Comment:
Same remark here and below: I don't think we should allow `T<:CatValue` in these constructors as that doesn't reflect the actual type which will be returned.

Reply:
But with the master it is possible:
```
julia> x = categorical(["a", "b", "c"])
3-element CategoricalArrays.CategoricalArray{String,1,UInt32}:
 "a"
 "b"
 "c"

julia> eltype(x)
CategoricalArrays.CategoricalValue{String,UInt32}

julia> CategoricalArray{eltype(x), 2}((2,2))
2×2 CategoricalArrays.CategoricalArray{String,2,UInt32}:
 #undef  #undef
 #undef  #undef
```

Reply:
Right, that comment is outdated given my reply above at https://github.com/JuliaData/CategoricalArrays.jl/pull/87#discussion_r145963843.